sklearn/utils/_unique.py

```diff
@@ -0,0 +1,45 @@
+# Authors: The scikit-learn developers
+# SPDX-License-Identifier: BSD-3-Clause
+
+import numpy as np
+from scipy.sparse import issparse
+
+from sklearn.utils._array_api import _is_numpy_namespace, get_namespace
+
+
+def attach_unique(y, xp=None):
+    """Attach unique values of y to y and return the result.
+
+    The result is a view of y, and the metadata (unique) is not attached to y.
+    """
+    xp, _ = get_namespace(y, xp=xp)
+    if not _is_numpy_namespace(xp) or issparse(y):
+        # we don't support non-numpy arrays here
+        return y
+
+    unique = xp.unique_values(y)
+    try:
+        unique_dtype = np.dtype(y.dtype, metadata={"unique": unique})
+        return y.view(dtype=unique_dtype)
```

Comment:
When you pickle or `np.save` this view, does the dtype metadata get serialized as well?

Reply:
pickle:

```py
In [1]: import numpy as np

In [2]: a = np.array([1])

In [4]: m_dtype = np.dtype(a.dtype, metadata={"unique": np.unique(a)})

In [5]: b = a.view(dtype=m_dtype)

In [7]: from pickle import loads, dumps

In [8]: loads(dumps(b)).dtype.metadata
Out[8]: mappingproxy({'unique': array([1])})
```

However, `np.save` complains:

```py
In [10]: from io import BytesIO

In [11]: f = BytesIO()

In [13]: np.save(f, b)
/home/adrin/micromamba/envs/sklearn/lib/python3.12/site-packages/numpy/lib/format.py:380: UserWarning: metadata on a dtype is not saved to an npy/npz. Use another format (such as pickle) to store it.
  d['descr'] = dtype_to_descr(array.dtype)
```

We're not gonna save these anyway, but interesting. cc @seberg

Reply:
Yes, we had to put this in because of h5py, you can't store metadata in npy format.

Reply:
We are not saving it, but a user may end up saving an ndarray with the attached metadata. In this case:

- If the unique values end up to be long, then it would increase their pickle size.
- If they use `np.save`, they will get a warning that they can ignore. (But it's hard to tell if it's safe to ignore)

Reply:
We are not modifying user's arrays. We create a new view and these arrays are not returned.

Reply:
Looking through the code now, I see that nothing is returned. Moving forward, we'll need to be careful that the output of `_check_targets` is not returned by a public function.

Reply:
That's a very good point, and I think it's a good idea to not modify `y` in any function that returns them. So I've moved `attach_unique` to the callers of `_check_targets` instead, so that in the future we don't have to worry about this point.